### PR TITLE
feat:porting over add adaptor/position/oracle/price router permission structure

### PR DESCRIPTION
**Pull request by cbrit**

### CODE REVIEW

File: proto/steward/v4/cellar_v2.proto [hunk 1]
```diff
@@ -46,37 +46,28 @@ import "curve.proto";
...
         // Represents function `setStrategistPayoutAddress(address payout)`
-        SetStrategistPayoutAddress set_strategist_payout_address = 5;
+        SetStrategistPayoutAddress set_strategist_payout_address = 4;
         // Represents function `swapPositions(uint256 index1, uint256 index2)`
-        SwapPositions swap_positions = 6;
+        SwapPositions swap_positions = 5;
         // Represents function `setShareLockPeriod(uint256 newLock)`
-        SetShareLockPeriod set_share_lock_period = 7;
-    }
+        SetShareLockPeriod set_share_lock_period = 6;
```
Comment (EricBolten) [inline proto/steward/v4/cellar_v2.proto:61]: I think we're missing some functions that we added on in 3.x-main as SP functions

`SetRebalanceDeviation`
`SetPlatformFee`
`InitiateShutdown`
`LiftShutdown`

File: proto/steward/v4/cellar_v2.proto [hunk 2]
```diff
@@ -231,6 +248,19 @@ message CellarV2_2 {
             SetStrategistPayoutAddress set_strategist_payout_address = 7;
             // Represents function `swapPositions(uint256 index1, uint256 index2)`
             SwapPositions swap_positions = 8;
+
```
Comment (EricBolten) [inline proto/steward/v4/cellar_v2.proto:292]: Here I believe we are also missing

`SetRebalanceDeviation`
`SetShareLockPeriod`
`InitiateShutdown`
`LiftShutdown`
`SetStrategistPlatformCut`

File: proto/steward/v4/cellar_v2.proto [hunk 3]
```diff
@@ -231,6 +248,19 @@ message CellarV2_2 {
...
+            // the new governance model is still in early adoption for emergencies.
+
+            // Represents function `cachePriceRouter(bool checkTotalAssets, uint16 allowableRange, address expectedPriceRouter)`
+            CachePriceRouter cache_price_router = 9;
+            // Represents function `addAdaptorToCatalogue(address adaptor)`
+            AddAdaptorToCatalogue add_adaptor_to_catalogue = 10;
+            // Represents function `addPositionToCatalogue(uint32 positionId)`
+            AddPositionToCatalogue add_position_to_catalogue = 11;
+            // Represents function `forcePositionOut(uint32 index, uint32 positionId, bool inDebtArray)`
+            ForcePositionOut force_position_out = 12;
```
Comment (EricBolten) [inline proto/steward/v4/cellar_v2.proto]: This should be a gov function I believe. We don't have any exception rules to deal with it.

File: proto/steward/v4/cellar_v2.proto [hunk 4]
```diff
@@ -495,6 +568,21 @@ message CellarV2_5 {
             SetAlternativeAssetData set_alternative_asset_data = 17;
             // Represents function `setDropAlternativeAssetData(ERC20 _alternativeAsset)`
             DropAlternativeAssetData drop_alternative_asset_data = 18; 
+
```
Comment (EricBolten) [inline proto/steward/v4/cellar_v2.proto:653]: I believe here we are also missing

`SetRebalanceDeviation`
`SetStrategistPlatformCut`
`IncreaseShareSupplyCap`

File: proto/steward/v4/cellar_v2.proto [hunk 5]
```diff
@@ -495,6 +568,21 @@ message CellarV2_5 {
...
+            // TEMPORARY
+            // These are governance functions, but will be allowed in a limited capacity for SPs while
+            // the new governance model is still in early adoption for emergencies.
+
+            // Represents function `setSharePriceOracle(uint256 _registryId, ERC4626SharePriceOracle _sharePriceOracle)`
+            SetSharePriceOracle set_share_price_oracle = 19;
+            // Represents function `cachePriceRouter(bool checkTotalAssets, uint16 allowableRange, address expectedPriceRouter)`
+            CachePriceRouter cache_price_router = 20;
+            // Represents function `forcePositionOut(uint32 index, uint32 positionId, bool inDebtArray)`
+            ForcePositionOut force_position_out = 21;
```
Comment (EricBolten) [inline proto/steward/v4/cellar_v2.proto]: This seems to be present in `3.x-main` but I think it's meant to be a gov function as per our conversations with Seven Seas.

File: src/cellars/cellar_v2.rs [hunk 6]
```diff
@@ -122,7 +130,7 @@ pub fn get_encoded_governance_call(
     proposal_id: u64,
 ) -> Result<Vec<u8>, Error> {
     match function {
-        InitiateShutdown(_) => {
+        GovernanceFunction::InitiateShutdown(_) => {
```
Comment (EricBolten) [inline src/cellars/cellar_v2.rs:189]: For the ones we are moving to the SP side as per the above comments, we could either leave copies here so they are callable or remove them.

File: src/cellars/cellar_v2.rs [hunk 7]
```diff
@@ -3,34 +3,32 @@
...
+pub fn get_encoded_call(function: Function, cellar_id: String) -> Result<Vec<u8>, Error> {
     get_call(function, cellar_id).map(|call| call.encode())
 }
 
-pub fn get_call(function: StrategyFunction, cellar_id: String) -> Result<CellarV2Calls, Error> {
+pub fn get_call(function: Function, cellar_id: String) -> Result<CellarV2Calls, Error> {
     match function {
-        AddPosition(params) => {
-            check_blocked_position(&params.position_id)?;
+        Function::AddPosition(params) => {
```
Comment (EricBolten) [inline src/cellars/cellar_v2.rs:35]: We need to add the checks we've ported here:

`validate_new_position(&cellar_id, params.position_id, &V2_0_PERMISSIONS)?;`

File: src/cellars.rs [hunk 8]
```diff
@@ -86,6 +224,113 @@ pub fn normalize_address(address: String) -> String {
...
+    Ok(())
+}
+
+pub fn validate_force_position_out(
+    cellar_id: &str,
+    index: u32,
+    position_id: u32,
+    in_debt_array: bool,
+) -> Result<(), Error> {
+    let cellar_id_normalized = normalize_address(cellar_id.to_string());
```
Comment (EricBolten) [inline src/cellars.rs]: We may actually be able to remove this as I assume this was a one-time fix, and we can move `ForcePositionOut` to gov.

**Review by EricBolten — COMMENTED** [review]
Some feedback.

File: proto/steward/v4/cellar_v2.proto (hunk 3 above)
Comment (EricBolten) [reply proto/steward/v4/cellar_v2.proto]: OK actually we do, but look at the later comment for discussion.

File: proto/steward/v4/cellar_v2.proto (hunk 5 above)
Comment (EricBolten) [reply proto/steward/v4/cellar_v2.proto]: See also later comment, I see we are using it.

File: proto/steward/v4/cellar_v2.proto (hunk 3 above)
Comment (cbrit) [reply proto/steward/v4/cellar_v2.proto]: 3.x has one valid call for Turbo SOMM; it seemed similar to other exceptions we are porting:

https://github.com/PeggyJV/steward/blob/6ccf3bd7b1affaa1eca2ac7c4875529e1d744236/steward/src/cellars.rs#L304-L322

File: src/cellars.rs (hunk 8 above)
Comment (cbrit) [reply src/cellars.rs]: Ah responded to your earlier comment before seeing this. Yes I think it was one-off. will fix.

File: src/cellars/cellar_v2.rs [hunk 9]
```diff
@@ -113,6 +112,62 @@ pub fn get_call(function: StrategyFunction, cellar_id: String) -> Result<CellarV
...
+        Function::SetStrategistPlatformCut(params) => {
+            log_cellar_call(
+                CELLAR_NAME,
+                &SetStrategistPlatformCutCall::function_name(),
+                &cellar_id,
+            );
+            let call = SetStrategistPlatformCutCall { cut: params.amount };
+            Ok(CellarV2Calls::SetStrategistPlatformCut(call))
+        }
+        Function::SetRebalanceDeviation(params) => {
```
Comment (EricBolten) [inline src/cellars/cellar_v2.rs:163]: `SetRebalanceDeviation` in Cellar v2.2 has some checks on the args, should we replicate them here?

**Review by EricBolten — COMMENTED** [review]
One last comment.

File: src/cellars/cellar_v2.rs (hunk 9 above)
Comment (cbrit) [reply src/cellars/cellar_v2.rs:163]: I thought I recalled something like that.

File: src/cellars/cellar_v2_5.rs [hunk 10]
```diff
@@ -222,6 +224,127 @@ pub fn get_encoded_function(call: FunctionCall, cellar_id: String) -> Result<Vec
...
+        }
+        Function::AddAdaptorToCatalogue(params) => {
+            validate_new_adaptor(&cellar_id, &params.adaptor, &V2_5_PERMISSIONS)?;
+            log_cellar_call(
+                CELLAR_NAME,
+                &AddAdaptorToCatalogueCall::function_name(),
+                &cellar_id,
+            );
+
+            if let Err(err) = check_blocked_adaptor(&params.adaptor) {
```
Comment (EricBolten) [inline src/cellars/cellar_v2_5.rs]: I think this `if` check made it in here by accident. `validate_new_adaptor` checks blocked adaptors and this references governance functions.

File: src/cellars/cellar_v2_5.rs [hunk 11]
```diff
@@ -222,6 +224,127 @@ pub fn get_encoded_function(call: FunctionCall, cellar_id: String) -> Result<Vec
...
+        }
+        Function::AddPositionToCatalogue(params) => {
+            validate_new_position(&cellar_id, params.position_id, &V2_5_PERMISSIONS)?;
+            log_cellar_call(
+                CELLAR_NAME,
+                &AddPositionToCatalogueCall::function_name(),
+                &cellar_id,
+            );
+
+            if let Err(err) = check_blocked_position(&params.position_id, &V2_5_PERMISSIONS) {
```
Comment (EricBolten) [inline src/cellars/cellar_v2_5.rs]: Also this is extraneous.

**Review by EricBolten — COMMENTED** [review]
Actually a couple more comments.

**Review by EricBolten — APPROVED** [review]
LGTM

**Pull request merged**